### ISSUE TITLE
Suspense triggered when changing the locale after upgrading to 1.1.1

### ISSUE THREAD

**Issue by hipdev** [QuiiBz/next-international#227]
**Describe the bug**
After updated to the latest version 1.1.1 when changing the locale the site is doing a fullpage reload with a white screen.

The only change I did from v 1.0.1 was adding the locale param:
```javascript
      <I18nProviderClient locale={params.locale} >
      ...
      </I18nProviderClient>
```

**To Reproduce**

**Expected behavior**
I don't want a full reload, just a change in the locales like before

**Screenshots**

**About (please complete the following information):**
 - next-international version 1.1.1
 - Next.js version 13.5.4

**Comment by QuiiBz** [comment]
I see that you're not using the `fallback` prop in `I18nProviderClient`, which is used by a `Suspense` boundary internally. 

To give more context, it's not a full-page reload, but rather this `Suspense` boundary that is triggered when changing the locale while the new one loads.

**Comment by gustaveWPM** [comment]
> I see that you're not using the `fallback` prop in `I18nProviderClient`, which is used by a `Suspense` boundary internally.
> 
> To give more context, it's not a full-page reload, but rather this `Suspense` boundary that is triggered when changing the locale while the new one loads.

Could it be possible to create an option to configure whether we want to load all the locales during initialization or lazily load them for the client-side part? (In the `createI18nClient` parameters).

**Comment by hipdev** [comment]
@QuiiBz ok, I put the fallback prop but it is not the expected functionality, now what it does is show that component when changing a language, but I want it to work as it was before, I don't want to show a component that replaces the entire page, just change the language 🤔

**Comment by QuiiBz** [comment]
Yeah I know that it doesn't "solve" this issue. The thing is that I have no idea how to:

1) keep lazy-loading the locale
2) allow Static Rendering
3) never show a loading state

all at once. The current solution uses the `use()` function from React, which unlocked 2) while keeping 1), but now triggers a Suspense boundary instead of keeping the old value when changing the locale. I'll ask on Twitter today but I've already been struggling for many hours on this problem.

**Comment by gustaveWPM** [comment]
> I've published `next-international@1.1.2-rc.0` (#234) which might fix the issue, but it's a very ugly hack. Let me know if that works for you.

Maybe another ugly hack would be to "Cache" the imported locales in a _Namespace_?

Something like:

```ts
namespace LocalesCache {
  export const localesCollection: Partial<Record<LanguageFlag, VocabObj>> = {};
}

function populateLocalesCache(language: LanguageFlag) {
  const localesElements = noIdea(); // * ... Compute...
  LocalesCache.localesCollection[language] = noIdea2(localesElements); // * ... Populate
}

function localesAccessor(language: LanguageFlag) {
  if (LocalesCache.localesCollection[language] === undefined) populateLocalesCache(language);
  return LocalesCache.localesCollection[language] as ???;
}

export const getLocale = (language: LanguageFlag): ??? =>
  localesAccessor(language);
```

In reality, I think it comes down to making a proxy?
(I don't know how to use `Proxy` in JS).

**Comment by hipdev** [comment]
@QuiiBz I tried with the next-international@1.1.2-rc.0 version but is not working, I'm still seeing the issue.

**Comment by QuiiBz** [comment]
We've had a conversation with Dan on X (Twitter), and it doesn't seem possible right now: https://twitter.com/dan_abramov/status/1711417357183098974

I've tried many many things, and none works for these [3 points mentioned above](https://github.com/QuiiBz/next-international/issues/227#issuecomment-1752484783). I'm kinda running out of ideas.

If anyone wants to help, here are the steps:
- Fork/clone the repo
- `pnpm i`
- `cd packages/international-types && pnpm build`
- `cd packages/next-international && pnpm build`
- `cd examples/next-app && pnpm dev`
- Add an artificial delay to the client-side locales imports, similar to https://github.com/QuiiBz/next-international/blob/1659894ced80bcfc78db22604350df9a0c49f994/examples/next-app/locales/client.ts#L8
- Navigate to `/en/client` and switch to FR / EN. The goal is to never show the "Loading..." (the suspense fallback) when switching locale and when refreshing the page.

**Comment by andelkocvjetkovic** [comment]
You could create server actions to change the locale and then pass them as props to the client component. This way, the ```tsx <I18nProviderClient fallback={...}></I18nProviderClient> ``` wouldn't be needed at all, and everything could be done in a single request trip."

**Comment by QuiiBz** [comment]
Thanks for the suggestion. However, I have a few concerns:

I don't get how a Server Action would help removing the need for `I18nProviderClient`. We still have to load the current locale on the client, not only when changing the locale but also on the first load. Additionally, Server Actions are in beta, and that would also mean we could no longer have a fully static output.

**Comment by andelkocvjetkovic** [comment]
The main reason why I have to use the ```I18nProviderClient``` is to change the locale state. I handle all translation in server components.

**Comment by gustaveWPM** [comment]
> We've had a conversation with Dan on X (Twitter), and it doesn't seem possible right now: https://twitter.com/dan_abramov/status/1711417357183098974
> 
> I've tried many many things, and none works for these [3 points mentioned above](https://github.com/QuiiBz/next-international/issues/227#issuecomment-1752484783). I'm kinda running out of ideas.
> 
> If anyone wants to help, here are the steps:
> 
> * Fork/clone the repo
> * `pnpm i`
> * `cd packages/international-types && pnpm build`
> * `cd packages/next-international && pnpm build`
> * `cd examples/next-app && pnpm dev`
> * Add an artificial delay to the client-side locales imports, similar to https://github.com/QuiiBz/next-international/blob/1659894ced80bcfc78db22604350df9a0c49f994/examples/next-app/locales/client.ts#L8
> * Navigate to `/en/client` and switch to FR / EN. The goal is to never show the "Loading..." (the suspense fallback) when switching locale and when refreshing the page.

I think I would have done it more like this.
I'm a bit too stupid to know how to really test properly if it works "Better". (Sorry.)
https://github.com/gustaveWPM/next-international/commit/7cdda37c7feec9a8e4dc11663894647f64c83466

**Comment by QuiiBz** [comment]
Thanks for sharing! Hovewer, this part will still trigger the component to suspend and show the fallback when changing to a locale that isn't in cache:

![Screenshot 2023-10-11 at 09 16 51](https://github.com/QuiiBz/next-international/assets/43268759/79c5288a-d322-401f-be20-4e51bf92a9e8)

**Comment by gustaveWPM** [comment]
> Thanks for sharing! Hovewer, this part will still trigger the component to suspend and show the fallback when changing to a locale that isn't in cache:
> 
> ![Screenshot 2023-10-11 at 09 16 51](https://user-images.githubusercontent.com/43268759/274175328-79c5288a-d322-401f-be20-4e51bf92a9e8.png)

Yes. Idk what to do. :/
Load and cache _all_ the locales as soon createI18nProviderClient is called?

Doesn't seem to be the right thing to do. :/

**Comment by gustaveWPM** [comment]
Well, I've listened to the "La chanson des bisous" song and wrote this.
https://github.com/gustaveWPM/next-international/commit/50e02bdc818e1f4f8dbd1e75d2c325db2a76e971

It seems to work properly. I let you check this.
I have only implemented this for the app router (no pages impl).

**Comment by QuiiBz** [comment]
I think that's actually a great idea to load the locale manually within `usechangeLocale`, and **then** navigate to the new route. I'll try later today in #234

**Comment by gustaveWPM** [comment]
> I think that's actually a great idea to load the locale manually within `usechangeLocale`, and **then** navigate to the new route. I'll try later today in #234

That's what I've done in the commit of the bisous.

**Comment by QuiiBz** [comment]
Yeah that's precisely what I was mentioning, great idea!

**Comment by gustaveWPM** [comment]
> Yeah that's precisely what I was mentioning, great idea!

I'd just forgotten an "else" in my spaghetti code. I have done an amend it just now.

**Comment by QuiiBz** [comment]
@gustaveWPM @hipdev @ApacheEx I've published `next-international@1.1.2-rc.1`, could you update and let me know if it works as expected? I've tried and it seems to work great on my side.

**Comment by gustaveWPM** [comment]
Well, I still have some "Too much" fallbacks when I use the next/prev buttons.
(I _thought_ it wasn't the case with my ugly code?)

It also seems to not perfectly work when we use those buttons + the `rewriteDefault` strategy.
But, even though I'm a very pessimistic person, I think we've made progress?

I'm looking forward to read hipdev again. :)

**Comment by gustaveWPM** [comment]
I just think that we could avoid a lot of calls to the "use" function.

It would be a good thing to do some caching here **too** rather than juggling between the map and use (until there's finally some caching with a call to "changeCurrentLocale"). Or maybe I'm wrong?

```ts
const clientLocale = localesCache.get(locale) ?? use(locales[locale as keyof typeof locales]()).default;```
https://github.com/QuiiBz/next-international/commit/98f40222b3a28a57ecc94d317d6abc057459d48b#diff-17f758f6c25222646825815d84f2fdb41479ed4cf0ca8ea861068333290bbd6bR15

**Comment by QuiiBz** [comment]
Just published `next-international@1.1.2-rc.2` that adds more caching as suggested. Navigating in the example, I'm no longer seeing any suspense fallback when refreshing, changing the locale or going backward/forward.

I'm waiting for more folks to test and confirm that it's fixed before publishing an official release.

**Comment by ApacheEx** [comment]
Hi guys, just tried the last two releases:

1) `next-international@1.1.2-rc.1` - I always see the "fallback" of Suspense when I land to "not-found" page. It's stuck and not resolving.
2)  `next-international@1.1.2-rc.2` - works as expected, I see fallback only once and then page content.


p.s the way of how you solved it looks much better.

**Comment by hipdev** [comment]
I tried `next-international@1.1.2-rc.1` and `next-international@1.1.2-rc.2`, I'm still seeing a white page reload 😭 

Just in case, here is my code:
```javascript
      ...
      <I18nProviderClient locale={params.locale} >
        <NextUIProvider>
          <Toaster />
          {children}
        </NextUIProvider>
      </I18nProviderClient>
      ...
```

**Comment by hipdev** [comment]
> > I tried next-international@1.1.2-rc.1 and next-international@1.1.2-rc.2, I'm still seeing a white page reload 😭
> 
> Could you use the `fallback` prop from `I18nProviderClient` to make sure it's coming from next-international? Otherwise, I think we'll need a minimal reproduction as the issue is fixed for 3 of us.

Ok, here is a minimal reproduction issue:
[https://github.com/hipdev/next-international-issue](https://github.com/hipdev/next-international-issue)
And the preview:
[https://next-international-issue.vercel.app/en](https://next-international-issue.vercel.app/en)

This is the expected behavior I want. Here is using next-international@1.0.1:
[https://www.uhh.club/](https://www.uhh.club/)

Now the issue with next-international@1.1.2-rc.2
[https://uhh-git-test-next-international-issue-juliancho.vercel.app/](https://uhh-git-test-next-international-issue-juliancho.vercel.app/)

Just in case, the language switcher is in the bottom right.

I tried with codesandbox but looks like the middleware is not supported, or maybe I'm missing something
[https://codesandbox.io/p/sandbox/goofy-roentgen-rch36n?file=/middleware.ts:17,3](https://codesandbox.io/p/sandbox/goofy-roentgen-rch36n?file=/middleware.ts:17,3)

Thank you for the hard work on this amazing library @QuiiBz , I hope this information helps.

**Comment by QuiiBz** [comment]
Thanks a lot for the reproduction and deployments. I'm unable to reproduce the issue every single time I switch the locale, only around once every 10-20. Do you also have this behavior?

https://github.com/QuiiBz/next-international/assets/43268759/a8ad14ad-98e1-4c6e-9113-cdeddf294b58

**Comment by hipdev** [comment]
> Thanks a lot for the reproduction and deployments. I'm unable to reproduce the issue every single time I switch the locale, only around once every 10-20. Do you also have this behavior?
> 
>  Screen.Recording.2023-10-11.at.20.19.56.mov

With the minimal reproduction example is like that, with my page is happening more often, maybe because it's with more components?

**Comment by gustaveWPM** [comment]
> Thanks a lot for the reproduction and deployments. I'm unable to reproduce the issue every single time I switch the locale, only around once every 10-20. Do you also have this behavior?
> 
>  Screen.Recording.2023-10-11.at.20.19.56.mov

I've tried setting ridiculously large vocabulary objects to stress test and see if it wasn't coming from potentially too systematic calls to "flatten locales" and it doesn't seem to be coming from there.

Personally, I can't reproduce the bug on my machine with this code. :/

https://github.com/gustaveWPM/next-international-issue/commit/a1c2840529a0c75815b9bc0a0a7ba7be6137bfa4

**Comment by QuiiBz** [comment]
I'll go ahead and close this initial issue, as the current behaviour is already much better than the previous release. I've published [next-international@1.1.2](https://github.com/QuiiBz/next-international/releases/tag/1.1.2) which includes #234 and more small fixes!

@hipdev feel free to open a new issue so we can take a more in-depth look at your specific problem.

**Comment by gustaveWPM** [comment]
Hello @hipdev,

I was thinking about your issue today (the one related to NextUI, which we didn't dig so far: https://uhh-git-test-next-international-issue-juliancho.vercel.app/).

Has it been resolved since then?

I was wondering if it might be due to perhaps setting the i18n provider a bit "too early." In that case, when the language changes, all its children are re-rendered, and maybe that's causing the problem?

Likewise, if we just go from:
```tsx
      <NextUIProvider>
        <I18nProviderClient locale={params.locale}>
          <Toaster />
          {children}
        </I18nProviderClient>
      </NextUIProvider>
```

To:
```tsx
      <I18nProviderClient locale={params.locale}>
        <NextUIProvider>
          <Toaster />
          {children}
        </NextUIProvider>
      </I18nProviderClient>
```

Does it change something?

I hope you don't mind me asking this question even though the issue is closed, as I might start using NextUI myself in a few days, so I'm curious.

If this doesn't solve the problem, I'd be interested in looking into it in more detail with another issue. :slightly_smiling_face:

**Comment by hipdev** [comment]
> Hello @hipdev,
> 
> I was thinking about your problem today (the one related to NextUI, which we didn't dig so far: https://uhh-git-test-next-international-issue-juliancho.vercel.app/).
> 
> Has it been resolved since then?
> 
> I was wondering if it might be due to your components tree, which maybe nest the `i18nProviderClient` a bit "too early". In that case, when the language changes, all its children are maybe re-rendered (including `NextUIProvider`, and then cause an "horrible" white-page flickering). Maybe that's what is causing your problem?
> 
> Likewise, if we just go from:
> 
> ```tsx
> <I18nProviderClient locale={params.locale}>
>   <NextUIProvider>
>     <Toaster />
>     {children}
>   </NextUIProvider>
> </I18nProviderClient>
> ```
> 
> (As written in [#227 (comment)](https://github.com/QuiiBz/next-international/issues/227#issuecomment-1757924579))
> 
> To:
> 
> ```tsx
> <NextUIProvider>
>   <I18nProviderClient locale={params.locale}>
>     <Toaster />
>     {children}
>   </I18nProviderClient>
> </NextUIProvider>
> ```
> 
> Does it change something?
> 
> I hope you don't mind that I'm asking this question even though the current issue is closed. As I might start using NextUI myself in a few days, I'm curious.
> 
> If this doesn't solve the problem, I'd be interested in looking into it in more detail, in a new issue. 🙂

Hey @gustaveWPM !, I tried it and still had the same issue. Finally I ended using  next-intl and the issue is solved.

**Comment by QuiiBz** [comment]
>  I tried it and still had the same issue. Finally I ended using next-intl and the issue is solved.

Sorry to hear that, this specific issue seemed to be fixed for many of us in the latest release. I guess it's too late now, but again, feel free to open an issue with more details of your setup.

**Comment by gustaveWPM** [comment]
> > I tried it and still had the same issue. Finally I ended using next-intl and the issue is solved.
> 
> Sorry to hear that, this specific issue seemed to be fixed for many of us in the latest release. I guess it's too late now, but again, feel free to open an issue with more details of your setup.

Hello!

I'm also seeing very slight unwanted apparitions of the _Suspense_, and I think more generous caching (with the _FlattenLocales_ call included during caching, as I mentioned earlier) might fix this.

**Comment by QuiiBz** [comment]
`flattenLocales` isn't related to any suspense boundary, there might be something else in your app that triggers it (e.g. data fetching)

**Comment by gustaveWPM** [comment]
> `flattenLocales` isn't related to any suspense boundary, there might be something else in your app that triggers it (e.g. data fetching)

Hmmm, I haven't done any data fetching for now.
I should try to clone again the lib and edit it with some of my horrible lines of code to see...
